pyspecProcScripts/calc_error.py

```diff
@@ -0,0 +1,131 @@
+import pyspecdata as psd
+from .simple_functions import select_pathway
+import numpy as np
+
+
+# {{{ Functions to calculate var and mean of nan-masked data
+#     Pulled from time_domain_noise.py example
+def masked_mean_multi(x, axis=None):
+    "Calculate the mean of nan-masked data on a 1D axis"
+    assert axis is not None
+
+    def masked_mean(x):
+        return np.mean(x[np.isfinite(x)])
+
+    return np.apply_along_axis(masked_mean, axis, x)
+
+
+def masked_var_multi(x, var_has_imag=False, axis=None):
+    "Calculates the variance along a 1D axis"
+    assert axis is not None
+
+    def masked_var(x):
+        if var_has_imag:  # take average of variance along real and imag
+            return np.var(x[np.isfinite(x)], ddov=1) / 2
+        else:
+            return np.var(x[np.isfinite(x)], ddof=1)
+
+    return np.apply_along_axis(masked_var, axis, x)
+
+
+# }}}
+def calc_error(
+    s,
+    frq_slice,
+    signal_pathway,
+    error_path,
+    direct="t2",
+    indirect="nScans",
+    use_mask=False,
+    fl=None,
+):
+    """Calculates the propagation of error for the given signal.
+
+    Before declaring the error_path,
+    look at an examples such as integration_w_error.py to see how to
+    decide which excluded pathways to take the error over.
+
+    Parameters
+    ==========
+    s: nddata
+        Full nddata set in the frequency domain for which the error is being
+        propagated.
+    frq_slice: tuple
+        Frequency range over which the signal resides.
+    signal_pathway:   dict
+                Dictionary of the path of the desired signal.
+    error_path: dict
+                Dictionary of all coherence pathways that are
+                not the signal pathway.
+
+                Before declaring the error_path,
+                look at an examples such as integration_w_error.py to
+                see how to decide which excluded pathways to take the
+                error over.
+    direct:     str
+                Direct axis.
+    indirect:   str
+                Indirect axis.
+    use_mask: boolean
+        If you would rather calculate the error excluding all the phase cycling
+        noise this should be set to true. Specifically this will apply a mask
+        over the signal pathway as well as over the integration slice in all
+        other pathways. When False, the error will only be calculated for the
+        pathways in error_path.
+
+    Returns
+    =======
+    collected_variance:       nddata
+             The error associated with coherence pathways not included in the
+             signal pathway.
+    """
+
+    s = s[direct:frq_slice]
+    df = s.get_ft_prop(direct, "df")
+    N = len(s[direct])
+    collected_variance = (
+        psd.ndshape(
+            [psd.ndshape(s)[indirect], len(error_path)], [indirect, "pathways"]
+        )
+        .alloc()
+        .setaxis("pathways", error_path)
+    )
```

Comment:
allocates an array in the shape of the indirect with the error pathways as an extra dim (used below)

Reply:
this should be obsolete now.